### PR TITLE
K8s/Folders: Allow recursive creation of DTO

### PR DESCRIPTION
**Pull request by suntala**
**What is this feature?**

Make the toDTO function recursive.

**Why do we need this feature?**

This is needed in order to support various endpoints such as get. 

**Who is this feature for?**

Grafana Search and Storage

**Which issue(s) does this PR fix?**:

Relates to https://github.com/grafana/search-and-storage-team/issues/143

**Special notes for your reviewer:**

We're going to add a test to check the exact responses for the different folder endpoints as part of a different PR. That test will check the functionality of toDTO indirectly.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/api/folder.go [hunk 1]
```diff
@@ -875,22 +870,29 @@ func (fk8s *folderK8sHandler) newToFolderDto(c *contextmodel.ReqContext, item un
...
+			// #TODO version doesn't seem to be used--confirm or set it properly
+			Version:       1,
+			AccessControl: acMetadata,
+			ParentUID:     fold.ParentUID,
+		}, nil
 	}
 
 	// no need to check view permission for the starting folder since it's already checked by the callers
-	folderDTO, err := toDTO(f, false)
+	folderDTO, err := toDTO(f, createdBy, false)
```
Comment (JohnnyQQQQ) [inline pkg/api/folder.go:895]: Suggestion: Perhaps we can add two consts for what false and true would mean in `toDTO`. So we end up with something like 

```golang
folderDTO, err := toDTO(f, createdBy, dontCheckCanView)
// or
folderDTO, err := toDTO(f, createdBy, checkCanView)
```

**Review by JohnnyQQQQ — COMMENTED** [review]
LGTM code wise, added a few non-blocking comments. Somebody with more context should give the final stamp 👍

File: pkg/api/folder.go (hunk 1 above)
Comment (suntala) [reply pkg/api/folder.go]: I like it. Thanks for the review! Applied changes for both suggestions here: [5f46dae](https://github.com/grafana/grafana/pull/96439/commits/5f46daeb912dd57fb3e0cbf50be610a99ab9738e)

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM, would be perfect if we could add one test for this new functionality.

File: pkg/api/folder.go [hunk 2]
```diff
@@ -917,19 +853,93 @@ func (fk8s *folderK8sHandler) newToFolderDto(c *contextmodel.ReqContext, item un
...
+		ff := folder.Folder{
 			UID:   uid,
-			OrgID: c.SignedInUser.GetOrgID(),
 			Title: v,
 			URL:   url,
-		})
+		}
+		parentDTO, err := fk8s.toDTO(c, &ff, "", checkCanView)
+		if err != nil {
+			// #TODO should we log this error?
```
Comment (filewalkwithme) [inline pkg/api/folder.go:863]: I think we should log the error here... the [(hs *HTTPServer) newToFolderDto](https://github.com/grafana/grafana/blob/927d0dddfe5a0b25ab5f8c35cd0aa3482d5d522a/pkg/api/folder.go#L442) is emitting a log message for this case

File: pkg/api/folder.go [hunk 3]
```diff
@@ -917,19 +853,93 @@ func (fk8s *folderK8sHandler) newToFolderDto(c *contextmodel.ReqContext, item un
...
+	// #TODO Is there a preexisting function we can use instead, something along the lines of UserIdentifier?
+	parts := strings.Split(rawIdentifier, ":")
+	if len(parts) < 2 {
+		return ""
+	}
+	return parts[1]
+}
+
+func (fk8s *folderK8sHandler) toDTO(c *contextmodel.ReqContext, fold *folder.Folder, createdBy string, checkCanView bool) (dtos.Folder, error) {
+	// #TODO revisit how/where we get orgID
```
Comment (filewalkwithme) [inline pkg/api/folder.go:885]: I think we can remove this comment
```suggestion
```

**Comment by filewalkwithme** [comment]
I do agree with @JohnnyQQQQ , the change looks good to me, but it would be nice if we could at a testcase ensuring that the Folder DTO rendering will work as expected for nested folders

**Comment by suntala** [comment]
Thank you both for your reviews. Normally, I would have gone ahead and left this PR open to include your suggestions, but I will be OOO early next week. Some of the other k8s client implementations (such as Update) will depend on this change so I think it best to merge now and add some e2e tests as a separate PR.

File: pkg/api/folder.go (hunk 3 above)
Comment (suntala) [reply pkg/api/folder.go:885]: We still need to figure out which of the parent fields to populate and how to do so.

File: pkg/api/folder.go (hunk 2 above)
Comment (suntala) [reply pkg/api/folder.go:863]: Good point. We need to add logging overall. I will add this as a separate task.

**Pull request merged**